app/src/main/java/org/p2p/wallet/home/ui/main/adapter/TokenAdapter.kt

```diff
@@ -44,32 +54,46 @@ class TokenAdapter(
         when (holder) {
-            is TokenViewHolder -> holder.onBind(data[position] as HomeElementItem.Shown, isZerosHidden)
-            is TokenHiddenViewHolder -> holder.onBind(data[position] as HomeElementItem.Hidden, isZerosHidden)
-            is TokenButtonViewHolder -> holder.onBind(data[position] as HomeElementItem.Action)
-            is BannersViewHolder -> holder.onBind(data[position] as HomeElementItem.Banners)
-            is HeaderViewHolder -> holder.onBind(data[position] as HomeElementItem.Title)
+            is TokenViewHolder -> holder.onBind(data[position] as Shown, isZerosHidden)
+            is TokenHiddenViewHolder -> holder.onBind(data[position] as Hidden, isZerosHidden)
+            is TokenButtonViewHolder -> holder.onBind(data[position] as Action)
+            is BannersViewHolder -> holder.onBind(data[position] as Banners)
+            is HeaderViewHolder -> holder.onBind(data[position] as Title)
         }
     }
 
-    private fun mapGroups(tokens: List<HomeElementItem>, state: VisibilityState): List<HomeElementItem> =
-        tokens
-            .groupBy { it is HomeElementItem.Hidden }
-            .map { (isHidden, list) -> mapTokensGroup(isHidden, list, state) }
-            .flatten()
-
-    private fun mapTokensGroup(
-        isHidden: Boolean,
-        list: List<HomeElementItem>,
-        state: VisibilityState
-    ) = list.toMutableList().apply {
-        if (isHidden) {
-            add(0, HomeElementItem.Action(state))
-        } else {
-            add(0, HomeElementItem.Title(R.string.home_tokens))
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: MutableList<Any>) {
+        val fields = (payloads.firstOrNull() as? Set<*>)?.filterIsInstance<String>()
+
+        if (fields.isNullOrEmpty()) {
+            onBindViewHolder(holder, position)
+            return
+        }
+
+        val item = data[position]
+        fields.forEach { field ->
+            when (field) {
+                DIFF_FIELD_TOKEN_BALANCE -> (holder as TokenViewHolder).bindBalance(item as Shown)
+                DIFF_FIELD_HIDDEN_TOKEN_BALANCE -> (holder as TokenHiddenViewHolder).bindBalance(item as Hidden)
+                DIFF_FIELD_TOGGLE_BUTTON -> (holder as TokenButtonViewHolder).bindIcon(item as Action)
+                DIFF_FIELD_TITLE -> (holder as HeaderViewHolder).onBind(item as Title)
+            }
         }
     }
 
+    private fun mapGroups(tokens: List<HomeElementItem>, state: VisibilityState): List<HomeElementItem> {
+        val result = mutableListOf<HomeElementItem>(Title(R.string.home_tokens))
+        val groups = tokens.groupBy { it is Hidden }
+        val hiddenTokens = groups[true].orEmpty()
+        val visibleTokens = groups[false].orEmpty()
+
+        result += visibleTokens
+        result += Action(state)
+        result += hiddenTokens
+
+        return result
+    }
+
     private fun getDiffCallback(
```

Comment:
let's extract this diff to a class? it's getting huge

Reply:
agree, nice point